### DISCUSSION TITLE
Help with checkbox props mismatching storybook controls

### DISCUSSION THREAD

**Discussion by acc-nicholas** [radix-ui/primitives#586] [Help] [answered]
Hello,

I'm a big fan of this project and am considering using for my company's design system. We are using storybook js to build it which i think you guys are familiar with since its part of some of the pkgs here.

Currently, i'm just trying to get a very simple `Checkbox` passthrough component working which is simply a styled version of the radix primitive checkbox.

However i'm running into something weird: i only see 5 controls/props in storybook, but the [radix docs](https://radix-ui.com/primitives/docs/components/checkbox#api-reference) show 9 props. The description is also missing.

My controls:
![image](https://user-images.githubusercontent.com/50001189/113623459-3e60ac00-9613-11eb-9d92-48a707706854.png)

My code:
```
// Checkbox.tsx
import * as RadixCheckbox from "@radix-ui/react-checkbox";

export const Checkbox: React.FC<RadixCheckbox.CheckboxOwnProps> = (props) => {
  return (
    <StyledCheckbox {...props}>
      <RadixCheckbox.Indicator as={CheckIcon} />
    </StyledCheckbox>
  );
};


// Checkbox.stories.tsx
import * as RadixCheckbox from "@radix-ui/react-checkbox";
const story = {
  title: "Components/Forms/Checkbox",
  component: Checkbox,
};
export default story as Meta;

const Template: Story<RadixCheckbox.CheckboxOwnProps> = (args) => (
  <Checkbox {...args} />
);
export const Default = Template.bind({});
```

It seems like `CheckboxOwnProps` is defined here:
https://github.com/radix-ui/primitives/blob/main/packages/react/checkbox/src/Checkbox.tsx#L21
and i can't seem to figure out where the other 4 props are inherited from. Not sure if this is a bug or if i'm doing something wrong.

Any help is much appreciated

**Comment by benoitgrelard** [comment]
Hi @acc-nicholas, thanks for your question, glad you like the components.

So in the docs I see:
- `as`
- `defaultChecked`
- `checked`
- `onCheckedChange`
- `disabled`
- `required`
- `readOnly`
- `name`
- `value`

and in your story I see:
- `defaultChecked`
- `checked`
- `onCheckedChange`
- `required`
- `readOnly`

I understand the confusion. I'll try and point out where the "missing" ones come from.
- `as` comes from the polymorphic types ([here](https://github.com/radix-ui/primitives/blob/main/packages/react/polymorphic/src/polymorphic.ts))
- `required` and `readOnly` are listed explicitly because we are adding them manually as we render a `button` and `button` DOM props doesn't have these
- the rest (`disabled`, `name`, `value`) are standard DOM props that are available on `button` which are automatically added by our polymorphic types

I hope that helps!

**Comment by benoitgrelard** [comment, marked as answer]
Oh I just realised also that the way you're setting the types on your component is not right.
I'm a bit confused as well because I see you referencing `RadixCheckbox.CheckboxOwnProps` but we don't export types so I'm wondering how that even works 🙂 

Regardless, as I was saying earlier, `CheckboxOwnProps` will only contain the props we've explicitly added to it.

If you don't care about polymorphism, you only need `React.elementRef` and `React.ComponentProps` (see in link below).
If you do need polymorphism, you can use our type helpers.
Have a look here for an example of both: https://radix-ui.com/primitives/docs/overview/styling#styling-a-part

I don't really know how the controls thing you're doing works, but looking at the code I'm guessing it's based on that template to which you give the props through the generic.
In this case I believe you would then be able to get most of the way by doing this:

```tsx
type Props = React.ComponentProps<typeof Checkbox>;
const Template: Story<Props> = (args) => (
  <Checkbox {...args} />
);
export const Default = Template.bind({});
```

I think the only issue you'll have with this is that the type of the `as` prop will probably be wrong/misleading. It'll probably say `button | undefined`.

**Reply by benoitgrelard** [reply]
cc @jjenzz is there another way you can think of? In particular at the end, instead of using `React.ComponentProps`?

**Reply by acc-nicholas** [reply]
Thanks for the response! it is very helpful!

Let me try what you have suggested. I think the storybook controls are auto-generated using the https://github.com/styleguidist/react-docgen-typescript pkg. I'm not too familiar with it, but it should take the prop type you specify and try to generate the controls automatically. But as you explained in this case, i'm missing some props because of the type i had found/chosen.

I will keep trying!

**Reply by acc-nicholas** [reply]
Ok some progress!

```
type CheckboxComponent = Polymorphic.ForwardRefComponent<
  Polymorphic.IntrinsicElement<typeof CheckboxPrimitive>,
  Polymorphic.OwnProps<typeof CheckboxPrimitive>
>;

export const Checkbox = React.forwardRef((props, forwardedRef) => {
  const themeContext = useContext(ThemeContext);
  return (
    <StyledCheckbox {...props}>
      <CheckboxPrimitive.Indicator
        as={CheckIcon}
        width={themeContext.space.medium}
        height={themeContext.space.medium}
      />
    </StyledCheckbox>
  );
}) as CheckboxComponent;
```
And the stories file:
```
type CheckboxProps = React.ComponentProps<typeof Checkbox>;
const Template: Story<CheckboxProps> = (args) => <Checkbox {...args} />;
export const Default = Template.bind({});
```

So this gets me `as` correctly as any html element and it picked up `ref` but i can hide this control manually.
![image](https://user-images.githubusercontent.com/50001189/114078526-e8337900-985d-11eb-97cd-b6892b1a0b88.png)

The only thing left is to get the html props that are automatically added by the polymorphic types. Any idea on how to pick those up?

**Reply by acc-nicholas** [reply]
Those html props depend on the `as` prop? So its very dynamic and we can't really know what props those may be?

**Reply by benoitgrelard** [reply]
Yeah at this point I'm not sure it's a limitation of that thing you're using then. I guess I'd be curious to see what it normally does with a component with no `as` prop but all the DOM props from a `button` for example. Would it list them here? Probably not right? Because that list would be insanely big and kinda useless. So I'm thinking they must have done something here to prevent some things to show regardless.

**Reply by jjenzz** [reply]
Yeah it looks to me like this tool only shows the custom props that are created and avoids listing all of the `JSX.IntrinsicElement` props which would be huge like Benoit says.

**Reply by jjenzz** [reply]
Having said that, I'm having a read through their [docs](https://github.com/styleguidist/react-docgen-typescript#propfilter) and it seems to suggest that it would document `React.FC<React.HTMLAttributes<HTMLDivElement>>` unless they have been removed via a `propFilter` configuration. 

> If you do not want to print out all the HTML attributes of a component typed like the following:
> ```
> const MyComponent: React.FC<React.HTMLAttributes<HTMLDivElement>> = ()...
> ```
> you can provide a propFilter function and do the filtering logic yourself.

Interesting 🤔  ... Do you have a `propFilter` that might be removing them?

**Reply by acc-nicholas** [reply]
> Would it list them here? Probably not right?

It appears it does not. I have something like that for my `Button` component:
```
export interface ButtonProps extends ButtonHTMLAttributes<HTMLElement> { ... }
```
And it doesn't show any of those html attributes in the prop/control table.

> Do you have a `propFilter` that might be removing them?

I do not have a prop filter afaik.

I think i'm okay with these html props not appearing. I can document and set this expectation globally.

Now i'm just ironing out any kinks... like the default `as` should be `button`, but does not appear. And the descriptions seem to be missing.

I did a quick test with `AccordionItem` and the descriptions came through with the same code, but not the default value for `as`.

**Reply by acc-nicholas** [reply]
I think `AccordionItem` has comments to describe the props here:
https://github.com/radix-ui/primitives/blob/main/packages/react/accordion/src/Accordion.tsx#L266

But checkbox is lacking those comments/descriptions here:
https://github.com/radix-ui/primitives/blob/main/packages/react/checkbox/src/Checkbox.tsx#L24

So that makes sense why they are missing for me. Shall I submit a feature request to add those comments/descriptions?

I'm still unsure why the default values are missing though. Will continue to look into that.

**Reply by benoitgrelard** [reply]
Yeah we haven't been consistent with adding descriptions like this in code because we currently have to do it manually for our docs so it's duplicated effort and hard to keep in sync.

In an ideal world, we would only describe every prop in code comments, and somehow use the types to generate the props in our docs but we're not quite there yet unfortunately.

**Reply by acc-nicholas** [reply]
Ah! I think the react-docgen-typescript expects default values to be defined in a certain way:

```
const Btn = ({
  variants = '',
  size = '',
  children,
  onClick,
  disabled = false,
  type = 'submit',
}: Props) => {
  /* ... */
};

```

But default values are defined differently in radix primitives:

```
const Checkbox = React.forwardRef((props, forwardedRef) => {
  const {
    as = CHECKBOX_DEFAULT_TAG,
    'aria-labelledby': ariaLabelledby,
    name,
    checked: checkedProp,
    defaultChecked,
    required,
    disabled,
    readOnly,
    value = 'on',
    onCheckedChange,
    ...checkboxProps
  } = props;
  ```
  
 Would you be open to changing this across all components?

**Reply by acc-nicholas** [reply]
> In an ideal world, we would only describe every prop in code comments, and somehow use the types to generate the props in our docs but we're not quite there yet unfortunately.

Seems like this is exactly what that `react-docgen-typescript` pkg is intended to do. It appears you are using storybook with story files. The doc + controls addons should be able to do this for you as well. So it seems like its possible to remove the manual work and de-dupe? Is storybook not working for you?

**Reply by benoitgrelard** [reply]
Not really, we don't use storybook for our documentation, we have a custom website for that.
